src/builtin.c

```diff
@@ -715,7 +715,7 @@ static jv f_format(jq_state *jq, jv input, jv fmt) {
     input = f_tostring(jq, input);
     const unsigned char* data = (const unsigned char*)jv_string_value(input);
     int len = jv_string_length_bytes(jv_copy(input));
-    size_t decoded_len = (3 * len) / 4; // 3 usable bytes for every 4 bytes of input
+    size_t decoded_len = (3 * (size_t)len) / 4; // 3 usable bytes for every 4 bytes of input
```

Comment:
I guess too late to change now, but is there a reason why `jv_*_length` etc don't return `size_t`?

Reply:
@wader I think it was done that was to make it fit in a `jv`; `jv`s are 128-bit values (so they can be returned by value without using the stack with `rax` and `rdx` in amd64), and they use 8-bit for flags, 8-bit for pad, 16-bit for offset, 32-bit for size (weirdly as sized `int` instead of unsigned), and 64-bit for either a pointer to an allocate value  or an embedded `double`. If both `size` and `u.ptr` were 64-bits (size `size_t`), you would not have space for flags, etc., and you could not efficiently return `jv`s by value, having to use something like `jv_get(&output, input, path)` instead of `output = jv_get(input, path)`.

Though, I am not sure why `size` has to be in `jv`, and not in the allocated value; I have never read that code.

Reply:
Huh interesting! i wonder much performance is gained by doing it that way? 🤔 gojq uses a `any` as value type which is two pointers https://github.com/golang/go/blob/bf86aec25972f3a100c3aa58a6abcbcc35bdea49/src/runtime/runtime2.go#L143-L146 are there any comprehensive benchmarks about this? jaq has some https://github.com/01mf02/jaq?tab=readme-ov-file#performance which seems to suggest jq, gojq and jaq are not that different, i would have expected more difference :)

Reply:
Two pointers are 2\*64-bit = 128-bit, so the same size as a `jv`.

What I meant to say is that with the C amd64 calling convention, that is the larger ("integer") value you can return efficiently by value just using registers; declaring a function that returns a larger value results in code similar to something like this:
```c
struct foo { size_t x[3]; }; /* 192-bit */
struct foo bar(int);
/* ... */
    struct foo v = bar(123);
/* automatically compiles to */
    struct foo v;
    (void)bar(&v, 123); /* returns &v */
```

Since it uses memory instead of just registers, it is slower and generally avoided to frequently return something by value when it is above 128-bit; so my guess was that `size` was made a 32-bit integer to fit in a `jv` that can be returned by value, with flags, value, etc.

(Page 27 (pdf) of https://cs61.seas.harvard.edu/site/pdf/x86-64-abi-20210928.pdf for formal specification of return by value convention)

Reply:
👍  (continued on discord)